[core-xml] Migrate to fast-xml-parser

joheredi opened the pull request:
Migrating core-xml dependency on xml2js -> fast-xml-parser.

**Motivation**
@azure/core-xml depends on `xml2js` for xml parsing and building. However this library is not being actively maintained and has potentially problematic issues such as https://github.com/Leonidas-from-XIV/node-xml2js/issues/593.

**Bundle impact**
Since fast-xml-parser is compatible with both browser and node we could go with this change in 2 ways 1) use the same implementation accross platforms or 2) just replace `xml2js` dependency and maintain the `.browser` implementation.

Going with (1) has the advantage of just keeping a single branch for parsing and building xml objects, but it comes with a bundle size penalty (2) saves us some bundle size but increases the maintenance cost.

In a quick analysis of a simple project that imports @azure/core-xml to consume parseXml switching to using FXP in the browser bundle adds 17kb (2kb -> 19kb)

**Changes**

The main difference between the libraries is how they represent attributes. xml2js has them in a child object with key "$" while fxp has them as siblings of the other nodes for example

**xml2js**
```json
// <fruit foo="true">Text</fruit>
{
   "fruit": {
      "$": "true",
      "_": "Text"
   }
}
```

**FXP**
```json
// <fruit foo="true">Text</fruit>
{
   "fruit": {
      "@_foo": true,
      "#Text": "Text"
   }
}
```

Most of the changes in this prototype are:

- Convert FXP representation to match xml2js
- Allow dropping the root node as xml2js allows

WIP:

- [ ] Support custom Attribute and Text Keys
- [ ] Cleanup flatten and group functions

jeremymeng commented:
Now that it's easy to plugin xml functionality in core v2. Maybe add this in a new package core-xml-fxp?

deyaaeldeen commented:
@jeremymeng I remember that @ramya-rao-a did not like increasing the number of core packages before so I will check with her first before adding one.

@joheredi do we have to worry about the parsing speed given the additional step of compiling from the new representation to the old one?

ramya-rao-a commented:
> Now that it's easy to plugin xml functionality in core v2. Maybe add this in a new package core-xml-fxp?

Why not replace xml2j in place in core-xml if fxp has benefits over xml2js?

xirzec commented:
I agree we can do this in place, though I think it's fair to bump the major version of `core-xml` to allow us to mindfully update SDKs who depend on it, in case there are issues with the new parser.

I'm also a little curious about how always pulling in `fast-xml-parser` affects bundle size for the browser scenario, now that we're always dragging in an extra library?

jeremymeng commented:
> Why not replace xml2j in place in core-xml if fxp has benefits over xml2js?

I was concerned about the feature parity between xml2js and fxp.  Having different ones allows different packages to chose between them in case fxp doesn't meet certain need.  I am not against in-place though.

jeremymeng commented:
|        | latest version | unpacked size |
|--------|----------------|---------------|
| xml2js | 0.4.23         | 43.5 kB       |
| fxp    | 3.20.0         | 78.3 kB       |

joheredi commented:
> I agree we can do this in place, though I think it's fair to bump the major version of `core-xml` to allow us to mindfully update SDKs who depend on it, in case there are issues with the new parser.
> 
> I'm also a little curious about how always pulling in `fast-xml-parser` affects bundle size for the browser scenario, now that we're always dragging in an extra library?

Since fast-xml-parser is compatible with both browser and node we could go with this change in 2 ways 1) use the same implementation accross platforms or 2) just replace xml2js dependency and maintain the .browser implementation.

Going with (1) has the advantage of just keeping a single branch for parsing and building xml objects, but it comes with a bundle size penalty (2) saves us some bundle size but increases the maintenance cost.

In a quick analysis of a simple project that imports @azure/core-xml to consume parseXml switching to using FXP in the browser bundle adds 17kb (2kb -> 19kb)

xirzec commented:
> In a quick analysis of a simple project that imports @azure/core-xml to consume parseXml switching to using FXP in the browser bundle adds 17kb (2kb -> 19kb)

I'm reluctant to add to our bundle size, but maybe since the cost of xml is only paid by the services that use it, this isn't so bad? @chradek @bterlson thoughts?

chradek commented:
How easy is it to allow someone to specify an xml parser implementation to use? If I recall, the existing XML parser in browsers won't work in web/service workers because the DOMParser isn't available in there, so there's some benefit to switching to just fast-xml-parser, but that's also a pretty big increase in bundle size which I'd rather avoid if we can.

joheredi commented:
> How easy is it to allow someone to specify an xml parser implementation to use? If I recall, the existing XML parser in browsers won't work in web/service workers because the DOMParser isn't available in there, so there's some benefit to switching to just fast-xml-parser, but that's also a pretty big increase in bundle size which I'd rather avoid if we can.

One option would be to export a couple of additional functions `lightParseXML` and `lightStringifyXml` which would pickup the browser implementation when bundling while the regular `parseXML` and `stringifyXML` would pickup FXP

joheredi commented:
@xirzec, @jeremymeng , @deyaaeldeen - I have cleaned up the changes from he first iteration and I think it looks much better. Here is what changed from last time:

- Figured out the options needed to avoid flattening and grouping as I had before. This will remove the complexity that @deyaaeldeen was concerned about.
- Brought back the browser implementation as otherwise, it would hit the bundle size.
- All tests should be passing now
- Added a bunch of new test coverage. And almost no changes to the existing tests

joheredi commented on sdk/core/core-xml/test/xml.spec.ts:
```diff
@@ -340,11 +438,11 @@ describe("XML serializer", function() {
+      const xml = stringifyXML(
         {
           fruit: {
-            apples: undefined
```
This is the only change I had to do to the existing tests. FXP removes a node if its value is `undefined` which I think aligns with the JS concept of `undefined`. xml2js and DOM would make it an empty node <foo />.

Are you aware of any implications this change may have?

deyaaeldeen reviewed (approved):
I wonder if the parser speed could be affected with the custom options and if that matters at all to the customers of our xml-using clients. Otherwise, I believe this is an important change, so we no longer use a vulnerable 3rd party package.

Thanks a lot for doing this!

jeremymeng commented on sdk/core/core-xml/src/xml.ts:
```diff
@@ -1,76 +1,52 @@
-  emptyTag: "",
-  cdata: false
-};
+import { XmlOptions } from "./xml.common";
```
merge import with above?

jeremymeng commented on sdk/core/core-xml/src/xml.ts:
```diff
@@ -79,21 +55,33 @@ export function stringifyXML(obj: unknown, opts: XmlOptions = {}): string {
+  const parsedXml = parse(unescapeHTML(str), getParserOptions(opts));
+
+  if (!opts.includeRoot) {
+    for (const key of Object.keys(parsedXml)) {
```
even `for` loop is used here, in practice, it would be the first and only property in `parsedXml`?

joheredi replied:
Yeah it will bail out right away

jeremymeng reviewed (approved):
Looks good! Minor comment.  I wish we could run storage/service bus tests but they are not migrated to core v2 yet.

xirzec reviewed (approved):
I have no concerns with this direction, but I wonder if we could do something hacky to get some test coverage with storage/service bus (e.g. regenerate with new codegen and hack their tests a little)

joheredi commented:
Thanks for the review @jeremymeng, @deyaaeldeen, and @xirzec. I'll do a quick perf test between before/after these changes to get a good view of the perf impact this change of dependency may have. And also will try to get some storage tests running at least locally to verify this.

I'll post my findings here before merging

joheredi commented:
> Thanks for the review @jeremymeng, @deyaaeldeen, and @xirzec. I'll do a quick perf test between before/after these changes to get a good view of the perf impact this change of dependency may have. And also will try to get some storage tests running at least locally to verify this.
> 
> I'll post my findings here before merging

Alright, so I was able to do a quick comparison parsing and stringifying a complex XML file 100x each. There was no noticeable difference in execution time.

I was also able to hack around a little bit and get a sanity call to the storage service going through these changes in core-xml.  The call was to the blob service `getProperties` endpoint and the response was serialized correctly the same way it does before the changes in this PR.

Unfortunately, the work to completely migrate Storage to corev2 is big and I wasn't able to port more operations in a reasonable timeframe for this PR. However I believe the new tests and the sanity check to the live service I was able to perform should give us enough confidence to go ahead with this change
